### PR TITLE
Stock Locations

### PR DESCRIPTION
**Pull request by atiftonat**

### CODE REVIEW

File: application/use_cases/stock_locations/create_stock_location/create_stock_location_interactor.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class CreateStockLocationInteractor(Interactor):
+
+    def __init__(self, persistence_context: IPersistenceContext):
+        self.persistence_context = persistence_context
+
+    async def execute_async(self, input_port: CreateStockLocationInputPort, output_port: ICreateStockLocationOutputPort):
+        stock_location = StockLocation( 
+            #how does this work, the class() takes base entity,
```
Comment (BenTalese) [inline application/use_cases/stock_locations/create_stock_location/create_stock_location_interactor.py:19]: chat on discord

File: application/use_cases/stock_locations/delete_stock_location/delete_stock_location_entity_existence_checker.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    import IDeleteStockLocationOutputPort
+
+class DeleteStockLocationEntityExistenceChecker(EntityExistenceChecker):
+    
+    def __init__(self, entity_existence_checker: IEntityExistenceChecker):
+        self.entity_existence_checker = entity_existence_checker
+
+    async def execute_async(self, input_port: DeleteStockLocationInputPort, output_port: IDeleteStockLocationOutputPort):
+        if not self.entity_existence_checker.does_entity_exist(StockLocation, input_port.stock_location_id):
+            # self.has_failures = True
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/delete_stock_location_entity_existence_checker.py:18]: chat on discord

File: application/use_cases/stock_locations/delete_stock_location/delete_stock_location_entity_existence_checker.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+            # ^ is this dynamically defining this variable in this instance?
+            
+            return output_port.stock_location_not_found_async(input_port.stock_location_id)
+            #^ do we need to await? nothing else async is going on
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/delete_stock_location_entity_existence_checker.py:22]: chat on discord

File: application/use_cases/stock_locations/delete_stock_location/delete_stock_location_input_port.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+from clapy import InputPort
+
+from domain.entities.base_entity import EntityID
+
+#is accessibility a thing for these classes?
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/delete_stock_location_input_port.py:5]: chat on discord

File: application/use_cases/stock_locations/delete_stock_location/delete_stock_location_input_port.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+class DeleteStockLocationInputPort(InputPort):
+    stock_location_id = EntityID
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/delete_stock_location_input_port.py:7]: Shouldn't be assignment

Chat on discord

File: application/use_cases/stock_locations/delete_stock_location/delete_stock_location_interactor.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        self.persistence_context = persistence_context
+
+    async def execute_async(self, input_port: DeleteStockLocationInputPort, output_port: IDeleteStockLocationOutputPort):
+        _StockLocation = self.persistence_context \
+            .get_entities(StockLocation) \
+            .first_by_id(input_port.stock_location_id)
+
+        self.persistence_context.remove(_StockLocation)
+
+        # await
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/delete_stock_location_interactor.py:22]: reminder to await

File: application/use_cases/stock_locations/delete_stock_location/idelete_stock_location_output_port.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    async def stock_location_deleted_async(self, stock_location_id: EntityID) -> None: # Do i want to send out the ID?
+        pass
+
+    @abstractmethod
+    async def stock_location_not_found_async(self, stock_location_id : EntityID) -> None:
+        pass
+
+# We should do the create response right from the start, i think it was created id in the header of the response?
+    
+# Should we include a version in the routes for extensibility, e.g., api/v1/stock
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/idelete_stock_location_output_port.py:19]: i don't mind doing it, so long as there's actually a benefit to doing it

File: application/use_cases/stock_locations/delete_stock_location/idelete_stock_location_output_port.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
+from abc import ABC, abstractmethod
+
+from clapy import IOutputPort
+
+from domain.entities.base_entity import EntityID
+
+class IDeleteStockLocationOutputPort(IOutputPort, ABC):
+
+    @abstractmethod
+    async def stock_location_deleted_async(self, stock_location_id: EntityID) -> None: # Do i want to send out the ID?
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/idelete_stock_location_output_port.py:10]: i have no idea what we want to come out of a delete yet, we may not need anything to come out

File: application/use_cases/stock_locations/delete_stock_location/idelete_stock_location_output_port.py [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+        pass
+
+    @abstractmethod
+    async def stock_location_not_found_async(self, stock_location_id : EntityID) -> None:
+        pass
+
+# We should do the create response right from the start, i think it was created id in the header of the response?
```
Comment (BenTalese) [inline application/use_cases/stock_locations/delete_stock_location/idelete_stock_location_output_port.py:17]: not sure what you're asking here

File: application/use_cases/stock_locations/update_stock_location/update_stock_location_input_port.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
+from clapy import InputPort
+
+from application.infrastructure.attribute_change_tracker import \
+    AttributeChangeTracker
+from domain.entities.base_entity import EntityID
+
+class UpdateStockLocationInputPort(InputPort):
+    description = AttributeChangeTracker[str] = AttributeChangeTracker[str] #How restrictive do we want to be? Can they provide nothing?
```
Comment (BenTalese) [inline application/use_cases/stock_locations/update_stock_location/update_stock_location_input_port.py:8]: Chat on discord (assignment, plus restrictiveness)

File: framework/api/routes/stock_locations/create_stock_location_command.py [hunk 11]
```diff
@@ -0,0 +1,7 @@
+from dataclasses import dataclass
+from uuid import UUID
+
+
+@dataclass
+class CreateStockLocationCommand():
```
Comment (BenTalese) [inline framework/api/routes/stock_locations/create_stock_location_command.py:6]: You don't need the brackets here

File: framework/api/startup.py [hunk 12]
```diff
@@ -50,6 +52,7 @@ def register_routers(app: Flask):
     app.register_blueprint(MERCHANT_ROUTER)
     app.register_blueprint(PRODUCT_ROUTER)
     app.register_blueprint(STOCK_ITEM_ROUTER)
+    app.register_blueprint(STOCK_LOCATION_ROUTER)
```
Comment (BenTalese) [inline framework/api/startup.py:55]: This should be automatic now I hope

File: interface_adaptors/infrastructure/configure_services.py [hunk 13]
```diff
@@ -4,11 +4,13 @@
...
 from interface_adaptors.controllers.stock_item_controller import StockItemController
+from interface_adaptors.controllers.stock_location_controller import StockLocationController
 from interface_adaptors.controllers.user_controller import UserController
 
 
 def configure_interface_adaptors_services(service_provider: DependencyInjectorServiceProvider) -> None:
     service_provider.register_service(providers.Factory, MerchantController)
     service_provider.register_service(providers.Factory, ProductController)
     service_provider.register_service(providers.Factory, StockItemController)
+    service_provider.register_service(providers.Factory, StockLocationController)
```
Comment (BenTalese) [inline interface_adaptors/infrastructure/configure_services.py:15]: This should be automatic now

**Review by BenTalese — COMMENTED** [review]
Left comments

File: application/use_cases/stock_locations/create_stock_location/create_stock_location_interactor.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
+from clapy import Interactor
```
[the remark this answers is not in the document]
Comment (BenTalese) [reply application/use_cases/stock_locations/create_stock_location/create_stock_location_interactor.py:1]: ben look into blank line in vsc

[the remark this answers is not in the document]
Comment (atiftonat) [reply application/use_cases/stock_locations/create_stock_location/create_stock_location_interactor.py:1]: https://stackoverflow.com/questions/44704968/visual-studio-code-insert-newline-at-the-end-of-files

File: interface_adaptors/infrastructure/configure_services.py (hunk 13 above)
Comment (atiftonat) [reply interface_adaptors/infrastructure/configure_services.py:15]: works, pull

File: framework/api/startup.py (hunk 12 above)
Comment (atiftonat) [reply framework/api/startup.py:55]: not working, remove when working

File: framework/api/infrastructure/middleware.py [hunk 15]
```diff
@@ -27,16 +28,42 @@ async def deserialise_web_request():
         _RequestData: dict = request.get_json()
 
         for _AttributeName, _AttributeType in get_type_hints(REQUEST_OBJECTS_BY_ENDPOINT[_RequestEndpoint]).items():
-            _Data = _RequestData[_AttributeName]
+            
+            _Data = _RequestData.get(_AttributeName)
```
Comment (BenTalese) [inline framework/api/infrastructure/middleware.py:32]: does .get() with just the key raise an exception if the key is not present?

File: framework/api/infrastructure/middleware.py [hunk 16]
```diff
@@ -27,16 +28,42 @@ async def deserialise_web_request():
...
+
+            if isinstance(_AttributeType(_Data), AttributeChangeTracker):
```
Comment (BenTalese) [inline framework/api/infrastructure/middleware.py:34]: should be able to use issubclass(), no need to construct

File: framework/api/infrastructure/middleware.py [hunk 17]
```diff
@@ -27,16 +28,42 @@ async def deserialise_web_request():
...
+
+        Args:
+            attribute_name (str): The name of the attribute 
+            request_data (dict): The request data which may require deserialisation for the value
+                of key 'attribute_name'. If not found
+                The deserialised attribute change tracker will be inserted or override the corresponding 
+                value for 'attribute_name'... STILL WRITING DOCUMENTATION
+    '''
+    
+    if attribute_name not in request_data.keys():
```
Comment (BenTalese) [inline framework/api/infrastructure/middleware.py:57]: would this read better as if, elif & else without the returns?

File: framework/api/service_collection_builder.py [hunk 18]
```diff
@@ -31,16 +31,16 @@ def build_service_provider(self):
...
-        for _Root, _Directories, _Files in os.walk("framework/api/routes"):
+        
+        for _Root, _Directories, _Files in os.walk(os.path.normpath("framework/api/routes")):
 
             DIR_EXCLUSIONS = [r"__pycache__"]
             FILE_EXCLUSIONS = [r".*__init__\.py", r"^.*(?<!\.py)$"]
             Common.apply_exclusion_filter(_Directories, DIR_EXCLUSIONS)
             Common.apply_exclusion_filter(_Files, FILE_EXCLUSIONS)
-
+
```
Comment (BenTalese) [inline framework/api/service_collection_builder.py:41]: seems you're leaving whitespace, there's extensions for removing on save (or VSC settings) and for highlighting

File: framework/api/infrastructure/middleware.py (hunk 15 above)
Comment (atiftonat) [reply framework/api/infrastructure/middleware.py:32]: It returns None in that scenario. Also tested this scenario.

@overload 
 def get(self, __key: _KT) -> _VT | None: ...
 
*KT = KeyType
*VT = ValueType

File: framework/api/infrastructure/middleware.py (hunk 17 above)
Comment (atiftonat) [reply framework/api/infrastructure/middleware.py:57]: Reads better & more succinct. Next commit will contain this change.

File: framework/api/service_collection_builder.py (hunk 18 above)
Comment (atiftonat) [reply framework/api/service_collection_builder.py:41]: I have incorporated the following settings in my settings.json file & removed the white spaces referred to by this comment. Hopefully that should reduce/remove this issue.

    "editor.renderWhitespace": "all",
    "files.trimTrailingWhitespace": true

File: requirements.txt [hunk 19]
```diff
@@ -0,0 +1,11 @@
+apscheduler==3.10.4
```
[the remark this answers is not in the document]
Comment (atiftonat) [reply requirements.txt:1]: I have just been maintaining a single file by overriding it.
So keep this one?

File: application/use_cases/stock_locations/update_stock_location/update_stock_location_interactor.py [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+        _StockLocation: StockLocation = self.persistence_context \
+            .get_entities(StockLocation) \
+            .first_by_id(input_port.stock_location_id)
+
+        if input_port.description.has_been_set:
+            _StockLocation.description = input_port.description.value
+
+        self.persistence_context.update(_StockLocation)
+
+        #dto, do we send out dtos usually? is there a benefit?
```
[the remark this answers is not in the document]
Comment (atiftonat) [reply application/use_cases/stock_locations/update_stock_location/update_stock_location_interactor.py:25]: decided i don't really see the benefit atm, can add ltr if required

File: requirements.txt (hunk 19 above)
[the remark this answers is not in the document]
Comment (BenTalese) [reply requirements.txt:1]: ye keep this, delete others

**Pull request closed without merging**